### ISSUE TITLE
Benchmarks on NIAH, ListOps (Curriculum)

### ISSUE THREAD

**Issue by Longcodedao** [LongDang-72/long-attention#12]
### 1. Needle In a Haystack

Version 1: 
I try to implement this dataset Needle In A Haystack (Google Colab) version

``` {python}
def generate_niah(batch_size, seq_len, vocab_size):
    key = torch.randint(10, vocab_size, (batch_size,))
    # Ensure noise doesn't contain the "trigger" tokens (3, 4, 5, 6)
    noise = torch.randint(10, vocab_size, (batch_size, seq_len-10))
    start = torch.tensor([3, 4]).expand(batch_size, -1)
    end = torch.tensor([5, 4, 3, 6]).expand(batch_size, -1)

    inputs = []
    for b in range(batch_size):
        seq = torch.cat([start[b], key[b].unsqueeze(0), noise[b], end[b]])
        inputs.append(seq)

    return torch.stack(inputs).to(device), key.to(device)


input_ids, labels = generate_niah(1, 128, 100)
print(f"Shape of the input: {input_ids.shape}")
print(f"Shape of the labels: {labels.shape}")
```

The output of that code is: 
```{python}
Shape of the input: torch.Size([1, 125])
Shape of the labels: torch.Size([1])
```

**Comments** 
* The actual sequence length does not equal to the desirable sequence length hyperparameter 
   Actual Input Length = Start(2) + Key(1) + Noise (Seq_len - 10) + End (4) = Seq_len - 3

* Positional Cheating:
   *  The structure is always: [START] [START] [KEY] ... meaning that the key always starts at the index 2 
   * Since Neural Networks are lazy optimizers, it just learn a positional shortcut : "Copying the embedding at the index 2 to the output => Hard-coded memorization
   *  **FIXED**: We should randomize the index of the key inserted in the haystack so that it forces the model to learn a semantic mechanism (Induction Head): "Scan the sequence -> Look for the specific `Key` token -> Copy the value immediately following it". Proves the model can actually search
   
* Problem "Loss in the Middle" Evidence:
   * Research from Stanford and UC Berkeley (Liu et al., 2023) proved that LLMs have a "U-shaped" performance curve. They are great at remembering the Start and End of a prompt, but terrible at the Middle (link: https://arxiv.org/pdf/2307.03172)
    * This version of NIAH dataset places the Key at the Start (index 2). Even weak models perform well here due to Primacy Bias.
    * **FIXED**: We have to randomly places the key (including the middle position). If the model succeed with this, you will have the mathematical proof that this does not suffer from this problem. 

My Implementation: 
This version is suited for the CausalLM in the HuggingFace format where the predict token is shift by 1 comparing to the input index.

To do that, we actually need a flag token to denote that the next token is the key.

```{python}
class NIAHDataset(Dataset):
    def __init__(self, size, min_len, max_len, vocab_size):
        self.size = size
        self.min_len = min_len
        self.max_len = max_len
        self.vocab_size = vocab_size
        
        # Special Tokens
        self.start_tokens = torch.tensor([3, 4])
        self.trigger_tokens = torch.tensor([5, 4, 3, 6])
        self.flag_token = torch.tensor([2]) 

    def __len__(self):
        return self.size

    def __getitem__(self, idx):
        # 1. Randomize Length for this specific sample
        # This prevents the model from overfitting to a specific horizon
        curr_len = np.random.randint(self.min_len, self.max_len + 1)
        
        # 2. Generate Key
        key = torch.randint(10, self.vocab_size, (1,))

        # 3. Calculate Noise
        # Overhead: Start(2) + Flag(1) + Key(1) + Trigger(4) + Target(1) = 9
        noise_len = curr_len - 9
        if noise_len < 0: noise_len = 0 # Safety clipping

        noise = torch.randint(10, self.vocab_size, (noise_len,))

        # 4. Insert Key Randomly
        insert_idx = torch.randint(0, noise_len + 1, (1,)).item()
        
        input_ids = torch.cat([
            self.start_tokens,
            noise[:insert_idx],
            self.flag_token,     # The Flag
            key,                 # The Needle
            noise[insert_idx:],
            self.trigger_tokens,
            key
        ])

        labels = input_ids.clone()
        labels[:-1] = -100 
        
        return {"input_ids": input_ids, "labels": labels}
```

This is the output of that 
```{python}
examples = train_dataset[0]
input_ids, labels = examples['input_ids'], examples['labels']

print(f"Shape of the input_ids: {input_ids.shape}")
print(f"Shape of the labels: {labels.shape}")
print("\n")
print(f"Input_ids: {input_ids}")
print(f"Labels: {labels}")
```
```{bash}
Shape of the input_ids: torch.Size([17])
Shape of the labels: torch.Size([17])


Input_ids: tensor([  3,   4, 987,   2, 246, 694, 712,  73, 654,  81, 297, 429,   5,   4,
          3,   6, 246])
Labels: tensor([-100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,
        -100, -100, -100, -100,  246])
```

Since the length of this dataset is randomly sampled, we need to pad the remaining tokens with pad_token is 0

```{python}
def pad_collate_fn(batch):
    # Extract inputs and labels
    inputs = [item['input_ids'] for item in batch]
    labels = [item['labels'] for item in batch]
    
    # Pad sequences to the longest in the batch (dynamic padding)
    # batch_first=True makes shape [Batch, Seq_Len]
    inputs_padded = rnn_utils.pad_sequence(inputs, batch_first=True, padding_value=0)
    labels_padded = rnn_utils.pad_sequence(labels, batch_first=True, padding_value=-100)
    
    return {
        "input_ids": inputs_padded,
        "labels": labels_padded
    }
```

Result of that 
```{python}
train_dataset = NIAHDataset(size=50000, min_len = 10, max_len = 30,  vocab_size=VOCAB_SIZE)

# 2. Create the DataLoader
train_loader = DataLoader(
    train_dataset, 
    batch_size=4, 
    shuffle=True, 
    num_workers=2,# Can use parallel loading now
    collate_fn=pad_collate_fn
)

iter_loader = next(iter(train_loader))
inputs_ids, labels = iter_loader['input_ids'], iter_loader['labels']

print(inputs_ids)
print(labels)
```

```{bash}
tensor([[  3,   4, 207, 941, 101, 193,  45, 283, 841, 749, 391,   2, 842, 429,
         567, 917,   5,   4,   3,   6, 842,   0,   0,   0],
        [  3,   4, 992, 742, 849, 868, 702, 898, 282, 363, 126,   2, 440, 705,
         159, 821, 706, 151, 818,   5,   4,   3,   6, 440],
        [  3,   4, 908,   2, 249,  14, 328,  61,  60, 987, 590,  25, 773, 952,
         483, 951, 646, 119,   5,   4,   3,   6, 249,   0],
        [  3,   4,  75,  87,   2, 920, 790, 693,   5,   4,   3,   6, 920,   0,
           0,   0,   0,   0,   0,   0,   0,   0,   0,   0]])
tensor([[-100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,
         -100, -100, -100, -100, -100, -100, -100, -100,  842, -100, -100, -100],
        [-100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,
         -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,  440],
        [-100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,
         -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,  249, -100],
        [-100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100,
          920, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100]])

```
